src/Features/Core/Portable/FileHeaders/AbstractFileHeaderDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.IO;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FileHeaders
+{
+    internal abstract class AbstractFileHeaderDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        protected AbstractFileHeaderDiagnosticAnalyzer(string language)
+            : base(
+                IDEDiagnosticIds.FileHeaderMismatch,
+                CodeStyleOptions.FileHeaderTemplate,
+                language,
+                new LocalizableResourceString(nameof(FeaturesResources.The_file_header_is_missing_or_not_located_at_the_top_of_the_file), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.A_source_file_is_missing_a_required_header), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            RoslynDebug.AssertNotNull(DescriptorId);
+
+            var invalidHeaderTitle = new LocalizableResourceString(nameof(FeaturesResources.The_file_header_does_not_match_the_required_text), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            var invalidHeaderMessage = new LocalizableResourceString(nameof(FeaturesResources.A_source_file_contains_a_header_that_does_not_match_the_required_text), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            InvalidHeaderDescriptor = CreateDescriptorWithId(DescriptorId, invalidHeaderTitle, invalidHeaderMessage);
+        }
+
+        internal DiagnosticDescriptor MissingHeaderDescriptor => Descriptor;
+
+        internal DiagnosticDescriptor InvalidHeaderDescriptor { get; }
+
+        protected abstract AbstractFileHeaderHelper FileHeaderHelper { get; }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxTreeWithoutSemanticsAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(HandleSyntaxTree);
+
+        public void HandleSyntaxTree(SyntaxTreeAnalysisContext context)
+        {
+            var tree = context.Tree;
+            var root = tree.GetRoot(context.CancellationToken);
+
+            // don't process empty files
+            if (root.FullSpan.IsEmpty)
+            {
+                return;
+            }
+
+            var options = context.Options.AnalyzerConfigOptionsProvider.GetOptions(tree);
```

Comment:
Please use `context.GetOption(CodeStyleOptions.FileHeaderTemplate)` this will do the exact same thing.

Reply:
The approach here avoids a sync-over-async path that would occur for `context.GetOption`.

Reply:
I can use `context.Options.TryGetEditorConfigOption` though.

Reply:
I don't think it is a good practice embedding the source of options into analyzers - basically the fact that the option comes from editorconfig or some other random option source should not matter to analyzers. The extensions have been added to explicitly to enable the analyzers to become agnostic of underlying options storage mechanism. When this analyzer moves to shared layer, it will automatically call the sync implementation that only looks at AnalyzerConfigOptions from the CodeStyle layer.